src/Confluent.Kafka/Producer.cs

```diff
@@ -787,35 +831,75 @@ internal Producer(ProducerBuilder<TKey, TValue> builder)
         {
             if (deliveryHandler != null && !enableDeliveryReports)
             {
-                throw new ArgumentException("A delivery handler was specified, but delivery reports are disabled.");
-            }
-
-            var keyBytes = (keySerializer != null)
-                ? keySerializer.Serialize(message.Key, true, message, topicPartition)
-                : asyncKeySerializer.SerializeAsync(message.Key, true, message, topicPartition)
-                    .ConfigureAwait(continueOnCapturedContext: false)
-                    .GetAwaiter()
-                    .GetResult();
-
-            var valBytes = (valueSerializer != null)
-                ? valueSerializer.Serialize(message.Value, false, message, topicPartition)
-                : asyncValueSerializer.SerializeAsync(message.Value, false, message, topicPartition)
-                    .ConfigureAwait(continueOnCapturedContext: false)
-                    .GetAwaiter()
-                    .GetResult();
-
-            ProduceImpl(
-                topicPartition.Topic,
-                valBytes, 0, valBytes == null ? 0 : valBytes.Length, 
-                keyBytes, 0, keyBytes == null ? 0 : keyBytes.Length, 
-                message.Timestamp, topicPartition.Partition, 
-                message.Headers, 
-                new TypedDeliveryHandlerShim_Action<TKey, TValue>(
+                throw new InvalidOperationException("A delivery handler was specified, but delivery reports are disabled.");
+            }
+
+            byte[] keyBytes;
```

Comment:
The `BeginProduce` method has also been revamped in this PR to provide uniformity across all types of errors. In this case though, exceptions are never thrown, all errors are reported via the callback. It's not completely without doubt that this is correct (this style of method is not idiomatic c#, so there's not much precedent), but I believe it'll be the most convenient in practice for analogous reasons to the async case.

Reply:
I don't think QueueFull should be raised through the delivery report since it voids any way for the application to maintain produce ordering.

Reply:
exceptionally good point

Reply:
I've changed this to throw `ProduceException<K,V>`'s which are `KafkaException`s with extra information. Also, did the same with k/v exceptions, since now they aren't the odd ones out.